File: DynamicPrograming/python/fib.py
from functools import lru_cache


def fib(n: int) -> int:
    @lru_cache(maxsize=None)
    def helper(k: int) -> int:
        if k in (0, 1):
            return k
        return helper(k - 1) + helper(k - 2)

    return helper(n)

File: DynamicPrograming/csharp/Fib.cs
using System.Collections.Generic;

namespace DynamicProgrammingSolutions;

public static class Fib
{
    public static int Calculate(int n)
    {
        var memo = new Dictionary<int, int>();
        return Helper(n, memo);
    }

    private static int Helper(int n, IDictionary<int, int> memo)
    {
        if (n == 0 || n == 1)
        {
            return n;
        }

        if (memo.TryGetValue(n, out var cached))
        {
            return cached;
        }

        memo[n] = Helper(n - 1, memo) + Helper(n - 2, memo);
        return memo[n];
    }
}
